contracts/PreOrder.sol

```diff
@@ -0,0 +1,161 @@
+        string memory name_,
+        string memory symbol_,
+        uint16 supply_
+    ) BatchNFT(name_, symbol_, supply_) {}
+
+    // start pre-order, only allowed by ower
+    function startPreOrder(uint256 amount, uint256 participants)
+        public
+        onlyOwner
+    {
+        // if (inPreOrder == true) revert PreOrderAlreadyStarted();
+        require(inPreOrder == false, "pre-order already started.");
+        // min contribution shall always be larger than 0
+        // if (amount <= 0) revert IncorrectMinAmount(amount);
+        require(amount > 0, "incorrect minimum amount.");
+
+        // if (participants <= 0) revert IncorrectParticipants(participants);
+        require(participants > 0, "incorrect number of participants.");
+
+        participantsAllowed = participants;
+        minAmount = amount;
+        inPreOrder = true;
+    }
+
+    // end pre-order, only allowed by ower
+    function endPreOrder() public onlyOwner {
+        // if (inPreOrder != true) revert PreOrderNotStarted();
+        require(inPreOrder == true, "pre-order has not been started.");
+
+        inPreOrder = false;
+    }
+
+    // place a pre-order
+    function preOrder() public payable {
+        // if (inPreOrder != true) revert PreOrderNotStarted();
+        require(inPreOrder == true, "pre-order has not been started.");
+        // if (_preOrdered[msg.sender] > 0) revert PreOrderExists(msg.sender);
+        require(
+            _preOrdered[msg.sender] <= 0,
+            "pre-order exists for this msg.sender."
+        );
+        // validation against the minimum contribution amount
+        // if (msg.value < minAmount) revert IncorrectMinAmount(msg.value);
+        require(msg.value >= minAmount, "incorrect minimum amount.");
+        // if (_nextPreOrder > participantsAllowed) revert MaxParticipantsReached(participantsAllowed);
+        require(
+            _nextPreOrder <= participantsAllowed,
+            "maximum participants reached."
+        );
+
+        // lets start the index from 1, since default uint in mapping is 0 in _preOrdered
+        if (_nextPreOrder == 0) _nextPreOrder = 1;
+        _preOrders[_nextPreOrder] = Participant({
+            sender: msg.sender,
+            amount: msg.value,
+            time: block.timestamp
+        });
+        _preOrdered[msg.sender] = _nextPreOrder;
+        _nextPreOrder += 1;
+        _preOrderAmountTotal += msg.value;
+    }
+
+    // check if an address has placed an order
+    function preOrderExist(address addr) public view virtual returns (bool) {
+        return _preOrdered[addr] > 0;
+    }
+
+    // return participant's pre-order detail
+    function preOrderGet(address addr)
+        public
+        view
+        virtual
+        returns (Participant memory)
+    {
+        Participant memory p = _preOrders[_preOrdered[addr]];
+        return p;
+    }
+
+    // return the number of pre-orders
+    function preOrderCount() public view virtual returns (uint256) {
```

Comment:
looks unused

Reply:
Will it be useful in the case that we want to the total number of pre-ordered participants?

Reply:
Or we could make `_nextPreOrder` public?